src/syntax/rules.rkt

```diff
@@ -52,50 +95,35 @@
 
 (define (*fp-safe-simplify-rules*)
   (reap [sow]
-    (for ([ruleset (*rulesets*)])
-      (match-define (list rules groups types) ruleset)
+    (for ([(_ ruleset) (in-dict (*rulesets*))])
+      (match-define (list rules groups _) ruleset)
       (when (and (ormap (curry flag-set? 'rules) groups)
                  (set-member? groups 'simplify)
                  (set-member? groups 'fp-safe)
                  (filter rule-ops-supported? rules))
         (for ([rule (in-list rules)])
           (sow rule))))))
 
-;; Rule struct
-
-(struct rule (name input output itypes otype)
-        ;; Input and output are patterns
-        ;; itypes is a mapping, variable name -> representation
-        ;; otype is a representation
-        #:methods gen:custom-write
-        [(define (write-proc rule port mode)
-           (fprintf port "#<rule ~a>" (rule-name rule)))])
-
 ;;
 ;;  Rule loading
 ;;
 
-(define-values (type-of-rule repr-of-rule)
-  (let () ; `let` not `begin` since these are expanded in different phases
-    (define ((type/repr-of-rule get-info name) input output ctx)
-      (let loop ([input input] [output output])
-        (cond [(list? input)    (if (equal? (car input) 'if)
-                                    ; special case for 'if'
-                                    ; return the 'type/repr-of-rule' of the ift branch
-                                    (loop (caddr input) output)
-                                    (get-info (car input) 'otype))]
-              [(list? output)   (if (equal? (car output) 'if)
-                                    ; special case for 'if'
-                                    ; return the 'type/repr-of-rule' of the ift branch
-                                    (loop input (caddr output))
-                                    (get-info (car output) 'otype))]
-              ;; fallback: if symbol, check the ctx for the type
-              [(symbol? input)  (dict-ref ctx input)]
-              [(symbol? output) (dict-ref ctx output)]
-              [else             (error name "could not compute type of rule ~a -> ~a"
-                                            input output)])))
-    (values (type/repr-of-rule operator-info 'type-of-rule)
-            (type/repr-of-rule impl-info 'repr-of-rule))))
+(define ((type/repr-of-rule op-info name) input output ctx)
+  (let loop ([input input] [output output])
+    (match* (input output)
+      ; first, try the input expression
+      ; special case for `if` expressions
+      [((list 'if _ ift _) _) (loop ift output)]
```

Comment:
I assume no real changes in functionality here, just rewrite & move `if` case out?

Reply:
Yep, not sure when I changed this honestly